File: cdm-sources/cdm/margin/schedule/functions/StandardizedScheduleFXSwapNotional.java
package cdm.margin.schedule.functions;

import cdm.base.math.NonNegativeQuantitySchedule;
import cdm.base.math.NonNegativeQuantitySchedule.NonNegativeQuantityScheduleBuilder;
import cdm.base.math.UnitType;
import cdm.base.math.metafields.FieldWithMetaNonNegativeQuantitySchedule;
import cdm.observable.asset.PriceQuantity;
import cdm.observable.asset.PriceSchedule;
import cdm.observable.asset.metafields.FieldWithMetaPriceSchedule;
import cdm.observable.asset.metafields.ReferenceWithMetaPriceSchedule;
import cdm.product.common.settlement.ResolvablePriceQuantity;
import cdm.product.template.SettlementPayout;
import cdm.product.template.TradeLot;
import com.google.inject.ImplementedBy;
import com.rosetta.model.lib.expression.CardinalityOperator;
import com.rosetta.model.lib.functions.ModelObjectValidator;
import com.rosetta.model.lib.functions.RosettaFunction;
import com.rosetta.model.lib.mapper.MapperC;
import com.rosetta.model.lib.mapper.MapperS;
import com.rosetta.model.metafields.FieldWithMetaString;
import java.util.Optional;
import javax.inject.Inject;

import static com.rosetta.model.lib.expression.ExpressionOperators.*;

@ImplementedBy(StandardizedScheduleFXSwapNotional.StandardizedScheduleFXSwapNotionalDefault.class)
public abstract class StandardizedScheduleFXSwapNotional implements RosettaFunction {
	
	@Inject protected ModelObjectValidator objectValidator;

	/**
	* @param farLeg 
	* @param tradeLot 
	* @return quantity 
	*/
	public NonNegativeQuantitySchedule evaluate(SettlementPayout farLeg, TradeLot tradeLot) {
		NonNegativeQuantitySchedule.NonNegativeQuantityScheduleBuilder quantityBuilder = doEvaluate(farLeg, tradeLot);
		
		final NonNegativeQuantitySchedule quantity;
		if (quantityBuilder == null) {
			quantity = null;
		} else {
			quantity = quantityBuilder.build();
			objectValidator.validate(NonNegativeQuantitySchedule.class, quantity);
		}
		
		return quantity;
	}

	protected abstract NonNegativeQuantitySchedule.NonNegativeQuantityScheduleBuilder doEvaluate(SettlementPayout farLeg, TradeLot tradeLot);

	protected abstract MapperS<? extends PriceQuantity> priceQuantity(SettlementPayout farLeg, TradeLot tradeLot);

	protected abstract MapperC<? extends FieldWithMetaNonNegativeQuantitySchedule> exchangedCurrencies(SettlementPayout farLeg, TradeLot tradeLot);

	protected abstract MapperS<? extends FieldWithMetaNonNegativeQuantitySchedule> extractedExchangedCurrency(SettlementPayout farLeg, TradeLot tradeLot);

	public static class StandardizedScheduleFXSwapNotionalDefault extends StandardizedScheduleFXSwapNotional {
		@Override
		protected NonNegativeQuantitySchedule.NonNegativeQuantityScheduleBuilder doEvaluate(SettlementPayout farLeg, TradeLot tradeLot) {
			NonNegativeQuantitySchedule.NonNegativeQuantityScheduleBuilder quantity = NonNegativeQuantitySchedule.builder();
			return assignOutput(quantity, farLeg, tradeLot);
		}
		
		protected NonNegativeQuantitySchedule.NonNegativeQuantityScheduleBuilder assignOutput(NonNegativeQuantitySchedule.NonNegativeQuantityScheduleBuilder quantity, SettlementPayout farLeg, TradeLot tradeLot) {
			final FieldWithMetaNonNegativeQuantitySchedule fieldWithMetaNonNegativeQuantitySchedule = extractedExchangedCurrency(farLeg, tradeLot).get();
			if (fieldWithMetaNonNegativeQuantitySchedule == null) {
				quantity = null;
			} else {
				quantity = toBuilder(fieldWithMetaNonNegativeQuantitySchedule.getValue());
			}
			
			return Optional.ofNullable(quantity)
				.map(o -> o.prune())
				.orElse(null);
		}
		
		@Override
		protected MapperS<? extends PriceQuantity> priceQuantity(SettlementPayout farLeg, TradeLot tradeLot) {
			final MapperC<PriceQuantity> thenArg = MapperS.of(tradeLot).<PriceQuantity>mapC("getPriceQuantity", _tradeLot -> _tradeLot.getPriceQuantity())
				.filterItemNullSafe(item -> areEqual(item.<FieldWithMetaPriceSchedule>mapC("getPrice", _priceQuantity -> _priceQuantity.getPrice()).<PriceSchedule>map("Type coercion", fieldWithMetaPriceSchedule -> fieldWithMetaPriceSchedule.getValue()), MapperS.of(farLeg).<ResolvablePriceQuantity>map("getPriceQuantity", settlementPayout -> settlementPayout.getPriceQuantity()).<ReferenceWithMetaPriceSchedule>mapC("getPriceSchedule", resolvablePriceQuantity -> resolvablePriceQuantity.getPriceSchedule()).<PriceSchedule>map("Type coercion", referenceWithMetaPriceSchedule -> referenceWithMetaPriceSchedule.getValue()), CardinalityOperator.All).get());
			return MapperS.of(thenArg.get());
		}
		
		@Override
		protected MapperC<? extends FieldWithMetaNonNegativeQuantitySchedule> exchangedCurrencies(SettlementPayout farLeg, TradeLot tradeLot) {
			return priceQuantity(farLeg, tradeLot).<FieldWithMetaNonNegativeQuantitySchedule>mapC("getQuantity", _priceQuantity -> _priceQuantity.getQuantity());
		}
		
		@Override
		protected MapperS<? extends FieldWithMetaNonNegativeQuantitySchedule> extractedExchangedCurrency(SettlementPayout farLeg, TradeLot tradeLot) {
			if (areEqual(exchangedCurrencies(farLeg, tradeLot).<NonNegativeQuantitySchedule>map("Type coercion", fieldWithMetaNonNegativeQuantitySchedule -> fieldWithMetaNonNegativeQuantitySchedule.getValue()).<UnitType>map("getUnit", nonNegativeQuantitySchedule -> nonNegativeQuantitySchedule.getUnit()).<FieldWithMetaString>map("getCurrency", unitType -> unitType.getCurrency()).<String>map("Type coercion", fieldWithMetaString -> fieldWithMetaString.getValue()), MapperS.of("USD"), CardinalityOperator.Any).getOrDefault(false)) {
				final MapperC<? extends FieldWithMetaNonNegativeQuantitySchedule> thenArg0 = exchangedCurrencies(farLeg, tradeLot)
					.filterItemNullSafe(item -> areEqual(item.<NonNegativeQuantitySchedule>map("Type coercion", fieldWithMetaNonNegativeQuantitySchedule -> fieldWithMetaNonNegativeQuantitySchedule == null ? null : fieldWithMetaNonNegativeQuantitySchedule.getValue()).<UnitType>map("getUnit", nonNegativeQuantitySchedule -> nonNegativeQuantitySchedule.getUnit()).<FieldWithMetaString>map("getCurrency", unitType -> unitType.getCurrency()).<String>map("Type coercion", fieldWithMetaString -> fieldWithMetaString == null ? null : fieldWithMetaString.getValue()), MapperS.of("USD"), CardinalityOperator.All).get());
				return MapperS.of(thenArg0.get());
			}
			if (areEqual(exchangedCurrencies(farLeg, tradeLot).<NonNegativeQuantitySchedule>map("Type coercion", fieldWithMetaNonNegativeQuantitySchedule -> fieldWithMetaNonNegativeQuantitySchedule.getValue()).<UnitType>map("getUnit", nonNegativeQuantitySchedule -> nonNegativeQuantitySchedule.getUnit()).<FieldWithMetaString>map("getCurrency", unitType -> unitType.getCurrency()).<String>map("Type coercion", fieldWithMetaString -> fieldWithMetaString.getValue()), MapperS.of("EUR"), CardinalityOperator.Any).getOrDefault(false)) {
				final MapperC<? extends FieldWithMetaNonNegativeQuantitySchedule> thenArg1 = exchangedCurrencies(farLeg, tradeLot)
					.filterItemNullSafe(item -> areEqual(item.<NonNegativeQuantitySchedule>map("Type coercion", fieldWithMetaNonNegativeQuantitySchedule -> fieldWithMetaNonNegativeQuantitySchedule == null ? null : fieldWithMetaNonNegativeQuantitySchedule.getValue()).<UnitType>map("getUnit", nonNegativeQuantitySchedule -> nonNegativeQuantitySchedule.getUnit()).<FieldWithMetaString>map("getCurrency", unitType -> unitType.getCurrency()).<String>map("Type coercion", fieldWithMetaString -> fieldWithMetaString == null ? null : fieldWithMetaString.getValue()), MapperS.of("EUR"), CardinalityOperator.All).get());
				return MapperS.of(thenArg1.get());
			}
			if (areEqual(exchangedCurrencies(farLeg, tradeLot).<NonNegativeQuantitySchedule>map("Type coercion", fieldWithMetaNonNegativeQuantitySchedule -> fieldWithMetaNonNegativeQuantitySchedule.getValue()).<UnitType>map("getUnit", nonNegativeQuantitySchedule -> nonNegativeQuantitySchedule.getUnit()).<FieldWithMetaString>map("getCurrency", unitType -> unitType.getCurrency()).<String>map("Type coercion", fieldWithMetaString -> fieldWithMetaString.getValue()), MapperS.of("JPY"), CardinalityOperator.Any).getOrDefault(false)) {
				final MapperC<? extends FieldWithMetaNonNegativeQuantitySchedule> thenArg2 = exchangedCurrencies(farLeg, tradeLot)
					.filterItemNullSafe(item -> areEqual(item.<NonNegativeQuantitySchedule>map("Type coercion", fieldWithMetaNonNegativeQuantitySchedule -> fieldWithMetaNonNegativeQuantitySchedule == null ? null : fieldWithMetaNonNegativeQuantitySchedule.getValue()).<UnitType>map("getUnit", nonNegativeQuantitySchedule -> nonNegativeQuantitySchedule.getUnit()).<FieldWithMetaString>map("getCurrency", unitType -> unitType.getCurrency()).<String>map("Type coercion", fieldWithMetaString -> fieldWithMetaString == null ? null : fieldWithMetaString.getValue()), MapperS.of("JPY"), CardinalityOperator.All).get());
				return MapperS.of(thenArg2.get());
			}
			if (areEqual(exchangedCurrencies(farLeg, tradeLot).<NonNegativeQuantitySchedule>map("Type coercion", fieldWithMetaNonNegativeQuantitySchedule -> fieldWithMetaNonNegativeQuantitySchedule.getValue()).<UnitType>map("getUnit", nonNegativeQuantitySchedule -> nonNegativeQuantitySchedule.getUnit()).<FieldWithMetaString>map("getCurrency", unitType -> unitType.getCurrency()).<String>map("Type coercion", fieldWithMetaString -> fieldWithMetaString.getValue()), MapperS.of("GBP"), CardinalityOperator.Any).getOrDefault(false)) {
				final MapperC<? extends FieldWithMetaNonNegativeQuantitySchedule> thenArg3 = exchangedCurrencies(farLeg, tradeLot)
					.filterItemNullSafe(item -> areEqual(item.<NonNegativeQuantitySchedule>map("Type coercion", fieldWithMetaNonNegativeQuantitySchedule -> fieldWithMetaNonNegativeQuantitySchedule == null ? null : fieldWithMetaNonNegativeQuantitySchedule.getValue()).<UnitType>map("getUnit", nonNegativeQuantitySchedule -> nonNegativeQuantitySchedule.getUnit()).<FieldWithMetaString>map("getCurrency", unitType -> unitType.getCurrency()).<String>map("Type coercion", fieldWithMetaString -> fieldWithMetaString == null ? null : fieldWithMetaString.getValue()), MapperS.of("GBP"), CardinalityOperator.All).get());
				return MapperS.of(thenArg3.get());
			}
			if (areEqual(exchangedCurrencies(farLeg, tradeLot).<NonNegativeQuantitySchedule>map("Type coercion", fieldWithMetaNonNegativeQuantitySchedule -> fieldWithMetaNonNegativeQuantitySchedule.getValue()).<UnitType>map("getUnit", nonNegativeQuantitySchedule -> nonNegativeQuantitySchedule.getUnit()).<FieldWithMetaString>map("getCurrency", unitType -> unitType.getCurrency()).<String>map("Type coercion", fieldWithMetaString -> fieldWithMetaString.getValue()), MapperS.of("CHF"), CardinalityOperator.Any).getOrDefault(false)) {
				final MapperC<? extends FieldWithMetaNonNegativeQuantitySchedule> thenArg4 = exchangedCurrencies(farLeg, tradeLot)
					.filterItemNullSafe(item -> areEqual(item.<NonNegativeQuantitySchedule>map("Type coercion", fieldWithMetaNonNegativeQuantitySchedule -> fieldWithMetaNonNegativeQuantitySchedule == null ? null : fieldWithMetaNonNegativeQuantitySchedule.getValue()).<UnitType>map("getUnit", nonNegativeQuantitySchedule -> nonNegativeQuantitySchedule.getUnit()).<FieldWithMetaString>map("getCurrency", unitType -> unitType.getCurrency()).<String>map("Type coercion", fieldWithMetaString -> fieldWithMetaString == null ? null : fieldWithMetaString.getValue()), MapperS.of("CHF"), CardinalityOperator.All).get());
				return MapperS.of(thenArg4.get());
			}
			return exchangedCurrencies(farLeg, tradeLot)
				.sort(item -> item.<NonNegativeQuantitySchedule>map("Type coercion", fieldWithMetaNonNegativeQuantitySchedule -> fieldWithMetaNonNegativeQuantitySchedule == null ? null : fieldWithMetaNonNegativeQuantitySchedule.getValue()).<UnitType>map("getUnit", nonNegativeQuantitySchedule -> nonNegativeQuantitySchedule.getUnit()).<FieldWithMetaString>map("getCurrency", unitType -> unitType.getCurrency()).<String>map("Type coercion", fieldWithMetaString -> fieldWithMetaString == null ? null : fieldWithMetaString.getValue()))
				.first();
		}
	}
}

File: cdm-sources/src/cdm/margin/schedule/functions/StandardizedScheduleFXSwapNotional.py
# pylint: disable=line-too-long, invalid-name, missing-function-docstring, missing-module-docstring, superfluous-parens
# pylint: disable=wrong-import-position, unused-import, unused-wildcard-import, wildcard-import, wrong-import-order, missing-class-docstring
from __future__ import annotations
import sys
import datetime
import inspect
from decimal import Decimal
from rosetta.runtime.utils import *
from rosetta.runtime.func_proxy import replaceable, create_module_attr_guardian
from cdm.product.template.SettlementPayout import SettlementPayout
from cdm.product.template.TradeLot import TradeLot
from cdm.base.math.NonNegativeQuantitySchedule import NonNegativeQuantitySchedule

__all__ = ['StandardizedScheduleFXSwapNotional']


@replaceable
def StandardizedScheduleFXSwapNotional(farLeg: SettlementPayout | None, tradeLot: TradeLot | None) -> NonNegativeQuantitySchedule:
    """
    Extracts the notional amount and currency of an FX swap.
    
    Parameters 
    ----------
    farLeg : SettlementPayout
    
    tradeLot : TradeLot
    
    Returns
    -------
    quantity : NonNegativeQuantitySchedule
    
    """
    self = inspect.currentframe()
    
    
    def _then_fn4():
        return (lambda item: get_only_element(item))(rosetta_filter(rosetta_resolve_attr(self, "exchangedCurrencies"), lambda item: all_elements(rosetta_resolve_attr(rosetta_resolve_attr(self, "unit"), "currency"), "=", "CHF")))
    
    def _else_fn4():
        return sorted(rosetta_resolve_attr(self, "exchangedCurrencies"))[0]
    
    def _then_fn3():
        return (lambda item: get_only_element(item))(rosetta_filter(rosetta_resolve_attr(self, "exchangedCurrencies"), lambda item: all_elements(rosetta_resolve_attr(rosetta_resolve_attr(self, "unit"), "currency"), "=", "GBP")))
    
    def _else_fn3():
        return if_cond_fn(all_elements(rosetta_resolve_attr(rosetta_resolve_attr(rosetta_resolve_attr(self, "exchangedCurrencies"), "unit"), "currency"), "=", "CHF"), _then_fn4, _else_fn4)
    
    def _then_fn2():
        return (lambda item: get_only_element(item))(rosetta_filter(rosetta_resolve_attr(self, "exchangedCurrencies"), lambda item: all_elements(rosetta_resolve_attr(rosetta_resolve_attr(self, "unit"), "currency"), "=", "JPY")))
    
    def _else_fn2():
        return if_cond_fn(all_elements(rosetta_resolve_attr(rosetta_resolve_attr(rosetta_resolve_attr(self, "exchangedCurrencies"), "unit"), "currency"), "=", "GBP"), _then_fn3, _else_fn3)
    
    def _then_fn1():
        return (lambda item: get_only_element(item))(rosetta_filter(rosetta_resolve_attr(self, "exchangedCurrencies"), lambda item: all_elements(rosetta_resolve_attr(rosetta_resolve_attr(self, "unit"), "currency"), "=", "EUR")))
    
    def _else_fn1():
        return if_cond_fn(all_elements(rosetta_resolve_attr(rosetta_resolve_attr(rosetta_resolve_attr(self, "exchangedCurrencies"), "unit"), "currency"), "=", "JPY"), _then_fn2, _else_fn2)
    
    def _then_fn0():
        return (lambda item: get_only_element(item))(rosetta_filter(rosetta_resolve_attr(self, "exchangedCurrencies"), lambda item: all_elements(rosetta_resolve_attr(rosetta_resolve_attr(self, "unit"), "currency"), "=", "USD")))
    
    def _else_fn0():
        return if_cond_fn(all_elements(rosetta_resolve_attr(rosetta_resolve_attr(rosetta_resolve_attr(self, "exchangedCurrencies"), "unit"), "currency"), "=", "EUR"), _then_fn1, _else_fn1)
    
    priceQuantity = (lambda item: get_only_element(item))(rosetta_filter(rosetta_resolve_attr(rosetta_resolve_attr(self, "tradeLot"), "priceQuantity"), lambda item: all_elements(rosetta_resolve_attr(item, "price"), "=", rosetta_resolve_attr(rosetta_resolve_attr(rosetta_resolve_attr(self, "farLeg"), "priceQuantity"), "priceSchedule"))))
    exchangedCurrencies = rosetta_resolve_attr(rosetta_resolve_attr(self, "priceQuantity"), "quantity")
    extractedExchangedCurrency = if_cond_fn(all_elements(rosetta_resolve_attr(rosetta_resolve_attr(rosetta_resolve_attr(self, "exchangedCurrencies"), "unit"), "currency"), "=", "USD"), _then_fn0, _else_fn0)
    quantity =  rosetta_resolve_attr(self, "extractedExchangedCurrency")
    
    
    return quantity

sys.modules[__name__].__class__ = create_module_attr_guardian(sys.modules[__name__].__class__)
